DateTime from varchar coercion.

fiedukow opened the pull request:
THIS IS CONTINUATION OF #128 

State of DateTime* and `VARCHAR` implicit cast after this task should be:
 * `VARCHAR` -> DateTime implicit cast works when there is no ambiguity.
 * `VARCHAR` -> DateTime casts will work with each type of DateTime format regardless of target type.
 * When there is ambiguity the @arhimondr algorithm for most specific function will be used.
 * The implicit cast from `VARCHAR` are allowed for the following target types: `TIMESTAMP`, `TIMESTAMP WITH TIME ZONE`, `TIME`, `TIME WITH TIME ZONE`, `DATE`

In examples:
 * `SELECT '2016-02-19' < DATE '2016-02-20';` will return `true`
 * `SELECT '2016-02-19 18:32 +01:00' < DATE '2016-02-20'; `will return `true`
 * `SELECT DAY('2016-02-19'); will return 19, but SELECT HOUR('2016-02-19 18:32');` will throw ambiguity due to the fact that the `VARCHAR` may be converted to `TIME` or `TIMESTAMP` and those types are not coertable.
 * `SELECT TIME '12:12' < TIMESTAMP '2016-02-19 18:32'` will not be executed due to lacking operator.

Those assumptions are covered in tests.

\* - by DateTime i mean all types of DATE, TIME, TIME WITH TIME ZONE, TIMESTAMP and TIMESTAMP WITH TIME ZONE.

losipiuk commented:
s/coersion/coercion

losipiuk commented on presto-main/src/main/java/com/facebook/presto/type/TimeOperators.java:
```diff
@@ -122,7 +122,7 @@ public static Slice castToSlice(ConnectorSession session, @SqlType(StandardTypes
     {
         try {
-            return parseTimeWithoutTimeZone(session.getTimeZoneKey(), value.toStringUtf8());
+            return parseTime(session.getTimeZoneKey(), value.toStringUtf8());
```
is that backward compatible?

fiedukow replied:
Yes. Parse time tries to parse time with timezone and if it fails it tries time without timezone.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/type/TimeWithTimeZoneOperators.java:
```diff
@@ -114,6 +117,18 @@ public static Slice castToSlice(@SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long
 
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
+    public static long castFromSlice(ConnectorSession session, @SqlType(StandardTypes.VARCHAR) Slice value)
```
I am somewhat afraid that by adding new cast operators we can run into problems that some SQL queries which run correctly before change, no will start failing with ambiguity related error.

Is that possible?

fiedukow replied:
I do not know about such possibility.
Implicit cast will work, as the TIME (without TZ) is most specific function for the pair (TIME, TIME_WITH_TIME_ZONE) and now it can parse string with timezone. This is the reason of the change here: https://github.com/Teradata/presto/pull/129/files#diff-61e5a60eb4040ce099bf4fe133374e6aL125

Explicit cast is explicit so there is no ambiguity AFAIU.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:
```diff
@@ -212,6 +212,13 @@ public static boolean canCoerce(List<? extends Type> actualTypes, List<Type> exp
         return true;
     }
 
+    private static List<String> getVarcharSubtypes()
```
make it constant

losipiuk commented on presto-main/src/main/java/com/facebook/presto/util/DateTimeUtils.java:
```diff
@@ -134,6 +137,33 @@ public static long parseTimestampWithoutTimeZone(TimeZoneKey timeZoneKey, String
+
+    public static long timestampWithTimeZoneToTimestampWithoutTimeZone(long value)
+    {
+        return unpackMillisUtc(value);
```
what is happening here?

losipiuk replied:
Ok - never mind. I thought that shift is to drop milliseconds. 
We just drop TZ information here.

fiedukow replied:
This code is only moved from operators inside TimestampWithTimeZoneOperators.java in order to be reused. Those are converters between different encoding of DateTime types on long variable.

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestDate.java:
```diff
@@ -193,4 +188,32 @@ public void testLeast()
+    public void dateCanBeImplicitlyCastedFromVarchar()
+            throws Exception
+    {
+        assertFunction("'1999-01-01' < DATE '2000-01-01'", BOOLEAN, true);
```
Can you use `999` instead of `1999` - so lexegraphical order differs from chronological order?

fiedukow replied:
Thanks for suggestion. I've made a separate testcase from that (for all DateTime types).

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestDate.java:
```diff
@@ -193,4 +188,32 @@ public void testLeast()
+        assertFunction("'1999-01-01' < DATE '2000-01-01'", BOOLEAN, true);
+        assertFunction("DATE '1999-01-01' < '2000-01-01'", BOOLEAN, true);
+
+        assertFunction("'2000-01-01' = DATE '2000-01-01'", BOOLEAN, true);
```
Same here - it would be nice to use two different string representations of same date.

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestDate.java:
```diff
@@ -193,4 +188,32 @@ public void testLeast()
+    }
+
+    @Test
+    public void dateCoertForCompareOperators()
```
`dateCoercionsForCompareOperators`

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestDate.java:
```diff
@@ -193,4 +188,32 @@ public void testLeast()
+    }
+
+    @Test
+    public void dateDoesCoertForMostSpecificFunction()
```
s/Coert/Coerce/

Add some explanation what is the intention of this test.

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimeWithTimeZone.java:
```diff
@@ -233,4 +228,82 @@ public void testCastToSlice()
     }
+
+    @Test
+    public void testCastFromSlice()
```
testCastFromVarchar; Slice is just internal implementation used by Varchar type.

fiedukow replied:
It's used around here though. That's why I followed this overwhelming convention ;-)
Should I change names of all such tests (testCastFromSlice and testCastToSlice in all DateTime tests)?

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimestamp.java:
```diff
@@ -283,4 +278,41 @@ public void testLeast()
+    }
+
+    @Test
+    public void timestampCoertForCompareOperators()
```
timestampCoercionsForCompareOperators

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimestamp.java:
```diff
@@ -283,4 +278,41 @@ public void testLeast()
+    }
+
+    @Test
+    public void timestampDoesNotCoertForAmbiguousFunctionCalls()
```
Coerce

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTime.java:
```diff
@@ -181,4 +175,29 @@ public void testCastFromSlice()
+    }
+
+    @Test
+    public void timeCoertForCompareOperators()
```
timeCoercionsForCompareOperators

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTime.java:
```diff
@@ -181,4 +175,29 @@ public void testCastFromSlice()
+    }
+
+    @Test
+    public void timeDoesNotCoertForAmbiguousFunctionCalls()
```
Coerce

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimeWithTimeZone.java:
```diff
@@ -233,4 +228,82 @@ public void testCastToSlice()
+    }
+
+    @Test
+    public void timeWithTimeZoneCoertForCompareOperators()
```
timeWithTImeZoneCoercionsForCompareOperators

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimeWithTimeZone.java:
```diff
@@ -233,4 +228,82 @@ public void testCastToSlice()
+    }
+
+    @Test
+    public void timeWithTimeZoneDoesNotCoertForAmbiguousFunctionCalls()
```
Coerce

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimestampWithTimeZone.java:
```diff
@@ -366,4 +362,91 @@ public void testLeast()
+    }
+
+    @Test
+    public void timestampWithTimeZoneCoertForCompareOperators()
```
Coercions

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimestampWithTimeZone.java:
```diff
@@ -366,4 +362,91 @@ public void testLeast()
+    }
+
+    @Test
+    public void timestampWithTimeZoneDoesNotCoertForAmbiguousFunctionCalls()
```
Coerce

losipiuk commented on presto-main/src/test/java/com/facebook/presto/type/TestTimestampWithTimeZone.java:
```diff
@@ -366,4 +362,91 @@ public void testLeast()
+    }
+
+    @Test
+    public void timestampWithTimeZoneDoesCoertForMostSpecificFunction()
```
Coerce

losipiuk commented:
looks good modulo some spelling

fiedukow commented:
General comment about naming test. I've renamed test you pointed out, but not in a fashion you've suggested.
One thing is that there was no consequence in using 'coercion' and 'implicit cast' - I've renamed all test to use 'implicit cast'.
Second thing are test like:
`timestampWithTimeZoneCoertForCompareOperators`
I would like to avoid naming them like "timestampWithTimeZoneCoersionForCompareOperators" (nor 'implicit cast' ofc) as this break form of sentence that express what do we want class to do (or to do with) instead of what are we testing.

fiedukow commented:
time_from_varchar_coersion_ambiguous_2 with final version was created based on decimal_coersion which ends this PR, continuation will be on facebook site.
